File: test_ban.js
const http = require('http');

function req(method, path, body, headers = {}) {
  return new Promise((resolve, reject) => {
    const opts = {
      hostname: '127.0.0.1',
      port: 3000,
      path: path,
      method: method,
      headers: headers
    };
    if (body) {
      opts.headers['Content-Type'] = 'application/json';
      body = JSON.stringify(body);
      opts.headers['Content-Length'] = Buffer.byteLength(body);
    }
    const r = http.request(opts, (res) => {
      let data = '';
      res.on('data', chunk => data += chunk);
      res.on('end', () => resolve({code: res.statusCode, body: data}));
    });
    r.on('error', reject);
    if (body) r.write(body);
    r.end();
  });
}

(async () => {
    console.log("Logging in");
    let res = await req("POST", "/api/login/email", {email: "superadmin@example.com", password: "superadmin_pass"});
    console.log(res);
    let token = JSON.parse(res.body).token;
    
    console.log("Fetching users");
    res = await req("GET", "/api/users", null, {"Authorization": token});
    console.log(res.code);
    let users = JSON.parse(res.body);
    console.log(users.map(u => u.username));
    
    let id = users.find(u => u.username === 'student').id;
    console.log("Banning id " + id);
    res = await req("POST", "/api/users/ban", {user_id: id, is_banned: true}, {"Authorization": token});
    console.log(res);
})();

File: test_ban.py
import urllib.request
import json
import ssl

def request(url, method="GET", data=None, headers={}):
    req = urllib.request.Request(url, method=method)
    for k, v in headers.items():
        req.add_header(k, v)
    if data:
        req.data = json.dumps(data).encode('utf-8')
        req.add_header("Content-Type", "application/json")
    try:
        ctx = ssl.create_default_context()
        ctx.check_hostname = False
        ctx.verify_mode = ssl.CERT_NONE
        resp = urllib.request.urlopen(req, context=ctx)
        return resp.read().decode('utf-8'), resp.status
    except urllib.error.HTTPError as e:
        return e.read().decode('utf-8'), e.code

print("Logging in...")
body, code = request("http://127.0.0.1:3000/api/login/email", "POST", {"email": "superadmin@example.com", "password": "superadmin_pass"})
if code != 200:
    print("Login error:", body)
    exit(1)

j = json.loads(body)
token = j.get("token")
print("Got token length:", len(token) if token else 0)

print("\nFetching users...")
body, code = request("http://127.0.0.1:3000/api/users", "GET", headers={"Authorization": token})
if code != 200:
    print("Users error:", body)
    exit(1)

users = json.loads(body)
print("Users:", [u['username'] for u in users])

student_id = next(u['id'] for u in users if u['username'] == 'student')

print("\nBanning user id", student_id)
body, code = request("http://127.0.0.1:3000/api/users/ban", "POST", {"user_id": student_id, "is_banned": True}, headers={"Authorization": token})
print("Ban response:", code, body)

